### PR TITLE
feat(idpe 17789): provide job to CompactionJobDoneSink (formerly known as PartitionDoneSink)

### PR DESCRIPTION
**Pull request by wiedld**
Part of https://github.com/influxdata/idpe/issues/17789

Renaming the mod is in another PR, otherwise the diff was not properly detected. (a.k.a. make it easier on code reviewers)

---

* refactor(idpe-17789): rename PartitionDoneSink to CompactionJobSink (3fdef7954)

* feat(idpe-17789): update CompactionJobDoneSink comments, and have it consume CompactionJob on record() (a8882c752)

* feat(idpe-17789): update MockCompactionJobDoneSink comments, local variables, tests, and record() signature (dca4b9687)

* feat(idpe-17789): update ErrorKindCompactionJobDoneSinkWrapper comments, local variables, tests, and record() signature (30a0dbcac)

* feat(idpe-17789): update LoggingCompactionJobDoneSinkWrapper comments, local variables, tests, and record() signature (7b7674ff6)

* feat(idpe-17789): update MetricsCompactionJobDoneSinkWrapper comments, local variables, tests, and record() signature (d6a20a343)

* feat(idpe-17789): update CompactionJobDoneSinkToScheduler comments, local variables, tests, and record() signature (c819c6c61)

* feat(idpe-17789): rename partition_done_sink in the components and driver, to be compaction_job_done_sink (425e0f1c3)

## Checklist
- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/components/compaction_job_done_sink/logging.rs [hunk 1]
```diff
@@ -1,102 +1,106 @@
...
+                );
             }
             Err(e) => {
                 // log compactor errors, classified by compactor ErrorKind
                 error!(
                     %e,
                     kind=e.classify().name(),
-                    partition_id = partition.get(),
+                    partition_id = job.partition_id.get(),
                     "Error while compacting partition",
```
Comment (wiedld) [inline compactor/src/components/compaction_job_done_sink/logging.rs:56]: Intentionally left this error message as an error per partition. 🤷🏼‍♀️

File: compactor/src/components/partition_done_sink/mod.rs [hunk 2]
```diff
@@ -14,29 +14,21 @@ pub mod metrics;
 pub mod mock;
 pub mod outcome;
 
-/// Records "partition is done" status for given partition.
+/// Records "compaction job is done" status for given partition.
 #[async_trait]
-pub trait PartitionDoneSink: Debug + Display + Send + Sync {
-    /// Record "partition is done" status for given partition.
+pub trait CompactionJobDoneSink: Debug + Display + Send + Sync {
```
Comment (wiedld) [inline compactor/src/components/partition_done_sink/mod.rs:19]: Note: this mod gets renamed in a [followup PR](https://github.com/influxdata/influxdb_iox/pull/8369), in order to make the diff register more cleanly (a.k.a. easier code review).

File: compactor/src/components/partition_done_sink/metrics.rs [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
-use data_types::PartitionId;
+use compactor_scheduler::CompactionJob;
 use metric::{Registry, U64Counter};
 
 use crate::error::{DynError, ErrorKind, ErrorKindExt};
 
-use super::PartitionDoneSink;
+use super::CompactionJobDoneSink;
 
 const METRIC_NAME_PARTITION_COMPLETE_COUNT: &str = "iox_compactor_partition_complete_count";
```
Comment (wiedld) [inline compactor/src/components/partition_done_sink/metrics.rs:11]: Intentionally left this metric name as unchanged.

File: compactor/src/components/partition_done_sink/mock.rs [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
-use super::{DynError, PartitionDoneSink};
+use super::{CompactionJobDoneSink, DynError};
 
-/// Mock for [`PartitionDoneSink`].
+/// Mock for [`CompactionJobDoneSink`].
 #[derive(Debug, Default)]
-pub struct MockPartitionDoneSink {
-    last: Mutex<HashMap<PartitionId, Result<(), String>>>,
+pub struct MockCompactionJobDoneSink {
+    last: Mutex<HashMap<CompactionJob, Result<(), String>>>,
```
Comment (wiedld) [inline compactor/src/components/partition_done_sink/mock.rs:11]: Use in hashmap required the derive Eq and Hash for the `CompactionJob`.

File: compactor/src/components/hardcoded.rs [hunk 5]
```diff
@@ -44,9 +44,9 @@ use super::{
     },
     parquet_files_sink::{dispatch::DispatchParquetFilesSink, ParquetFilesSink},
     partition_done_sink::{
```
Comment (wiedld) [inline compactor/src/components/hardcoded.rs:46]: Note: this mod `partition_done_sink` gets renamed in a followup PR. To make for cleaner/easier code reviews.

File: compactor/src/components/mod.rs [hunk 6]
```diff
@@ -59,7 +59,7 @@ pub struct Components {
     /// condition to avoid running out of resources during compaction
     pub post_classification_partition_filter: Arc<dyn PostClassificationPartitionFilter>,
     /// Records "partition is done" status for given partition.
```
Comment (carols10cents) [inline compactor/src/components/mod.rs:61]: I think this comment needs to be updated?

**Review by carols10cents — APPROVED** [review]
Just one little comment that doesn't make as much sense anymore, otherwise looks great!

File: compactor/src/components/mod.rs (hunk 6 above)
Comment (wiedld) [reply compactor/src/components/mod.rs]: Done

**Pull request merged**